### PR TITLE
Disallow initialising from an iterable containing duplicates

### PR DESCRIPTION
**Pull request by Yogi-B**
Add `Ordering.assert_unique_items` class method to validate an iterable used to initialise an Ordering contains no duplicates, raising a ValueError otherwise.

Add unit test for rejecting duplicates at initialization.

### CODE REVIEW

File: ordering/__init__.py [hunk 1]
```diff
@@ -103,6 +104,14 @@ def assert_new_item(self, item: T) -> None:
         if item in self:
             raise KeyError("Ordering {} already contains {}".format(self, item))
 
+    @classmethod
+    def assert_unique_items(cls, items: Collection[_T]) -> None:
+        if len(items) > len(set(items)):
```
Comment (madman-bob) [inline ordering/__init__.py:109]: Could turn this into an early return

File: ordering/__init__.py [hunk 2]
```diff
@@ -103,6 +104,14 @@ def assert_new_item(self, item: T) -> None:
...
+            d: Dict[_T, bool] = {}
+            duplicates = [i for i in items if i in d or d.setdefault(i, False)]
```
Comment (madman-bob) [inline ordering/__init__.py:111]: This line feels a bit too much like magic for my liking

File: ordering/__init__.py [hunk 3]
```diff
@@ -103,6 +104,14 @@ def assert_new_item(self, item: T) -> None:
...
+            raise ValueError('attempt to create {} containing duplicate items: {}'
```
Comment (madman-bob) [inline ordering/__init__.py:112]: I use [this](https://stackoverflow.com/a/56190) convention for single/double quotes.

Comment (madman-bob) [reply ordering/__init__.py:112]: Capitalise the first letter of the error message

**Review by madman-bob — CHANGES_REQUESTED** [review]

File: ordering/__init__.py (hunk 1 above)
Comment (Yogi-B) [reply ordering/__init__.py:109]: I agree, though I was just following the other `assert_*` methods' if-condition-raise structure.

File: ordering/__init__.py (hunk 2 above)
Comment (Yogi-B) [reply ordering/__init__.py:111]: Fair enough, I'm also wondering if it's worth getting *all* the duplicate items, or just returning the first duplicate, à la overlapping positional and keyword arguments, i.e.
```Python
>>> (lambda x, y: None)(1, 2, x=1, y=2)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: <lambda>() got multiple values for argument 'x'
```

File: ordering/__init__.py (hunk 3 above)
Comment (Yogi-B) [reply ordering/__init__.py:112]: -  Personally I tend follow how the repr of the string looks, i.e. primarily using single quotes, then switching to double quotes when the string contains single quotes, finally switching back to single quotes if it contains both. However, I can see the logic in the convention you linked.<br><br>Also I tend to use triple single quotes for multiline strings that are not docstrings as depending on what editor I'm using it can allow for quickly differentiating between docstrings and multiline strings by their highlighting, even when the highlighter isn't directly aware of the concept of a docstring.

-  Okay, though I blame my non-capitalisation on the fact that unlike those in the standard library, most built-in error messages aren't capitalised.

File: ordering/__init__.py (hunk 1 above)
Comment (madman-bob) [reply ordering/__init__.py:109]: Fair. It just happened that the other `assert`s were shorter without the early return.

File: ordering/__init__.py (hunk 2 above)
Comment (madman-bob) [reply ordering/__init__.py:111]: I'm happy to go either way on that point.

Showing all the duplicate items may be more convenient, as it stops the potential cycle of back and forth of fixing one problem, just for another to pop up.

At the same time, it is possible that showing all duplicates may be overwhelming, and it doesn't match with the standard library.

Do whichever is most convenient for you.
